### DISCUSSION TITLE
`gevent.exceptions.InvalidSwitchError` in SQLAlchemy `_finalize_fairy`

### DISCUSSION THREAD

**Discussion by andrii-korotkov-verkada** [sqlalchemy/sqlalchemy#11866] [Usage Questions] [answered]
Hey. I'm very occasionally getting the request hang and stack trace looks like
```
Traceback (most recent call last):
  File "/site-packages/sqlalchemy/pool/base.py", line 763, in _finalize_fairy
    fairy._reset(pool, transaction_was_reset)
  File "/site-packages/sqlalchemy/pool/base.py", line 1038, in _reset
    pool._dialect.do_rollback(self)
  File "/site-packages/sqlalchemy/engine/default.py", line 683, in do_rollback
    dbapi_connection.rollback()
  File "/site-packages/psycogreen/gevent.py", line 36, in gevent_wait_callback
    wait_read(conn.fileno(), timeout=timeout)
  File "src/gevent/_hub_primitives.py", line 353, in gevent._gevent_c_hub_primitives.wait_read
  File "src/gevent/_hub_primitives.py", line 367, in gevent._gevent_c_hub_primitives.wait_read
  File "src/gevent/_hub_primitives.py", line 350, in gevent._gevent_c_hub_primitives.wait_on_watcher
  File "src/gevent/_hub_primitives.py", line 304, in gevent._gevent_c_hub_primitives._primitive_wait
  File "src/gevent/_hub_primitives.py", line 46, in gevent._gevent_c_hub_primitives.WaitOperationsGreenlet.wait
  File "src/gevent/_hub_primitives.py", line 46, in gevent._gevent_c_hub_primitives.WaitOperationsGreenlet.wait
  File "src/gevent/_hub_primitives.py", line 57, in gevent._gevent_c_hub_primitives.WaitOperationsGreenlet.wait
gevent.exceptions.InvalidSwitchError: Invalid switch into <Greenlet at 0x7effd9da1ae0: _handle_and_close_when_done(functools.partial(<bound method GeventWorker.handl, <bound method StreamServer.do_close of <StreamServ, (<gevent._socket3.socket at 0x7effd9dc40a0 object,)>: got None (expected <gevent._gevent_c_waiter.Waiter object at 0x7effd1ed9bd0>; waiting on <Hub '' at 0x7f0032c997c0 epoll default pending=0 ref=114 fileno=8 resolver=<gevent.resolver.thread.Resolver at 0x7f0024179760 pool=<ThreadPool at 0x7f002420d660 tasks=0 size=1 maxsize=10 hub=<Hub at 0x7f0032c997c0 thread_ident=0x7f0034181740>>> threadpool=<ThreadPool at 0x7f002420d660 tasks=0 size=1 maxsize=10 hub=<Hub at 0x7f0032c997c0 thread_ident=0x7f0034181740>> thread_ident=0x7f0034181740> with <io at 0x7effd9823840 native=0x7effd9823880 fd=96 events=READ active callback=<bound method Waiter.switch of <gevent._gevent_c_waiter.Waiter object at 0x7effd1ed9bd0>> args=(<gevent._gevent_c_waiter.Waiter object at 0x7effd1ed9bd0>,)>)
```
The issue is hard to reproduce and only happens on ~10% of the integration test runs of a very large test set, and running subsets of tests often don't reproduce the issue. So I don't have a good minimal repro.
I've only noticed that the issue seems to be gone after removing `psycogreen.gevent.patch_psycopg()`, which is called very early in the app __init__.
Library versions:
```
gevent==24.2.1
greenlet==3.0.3
gunicorn==23.0.0
sqlalchemy==1.4.51 or 1.4.54
sqlalchemy-utils==0.38.3 or 0.41.2
psycogreen==1.0.2
psycopg==3.1.19
psycopg2_binary==2.9.9
psycopg2_pool==1.2
```
This looks similar to https://github.com/sqlalchemy/sqlalchemy/discussions/9469, but the error is different.
Thanks for any help.

**Comment by zzzeek** [comment]
are the errors interruptive in the main thread that's running or do you just see the errors in the logs, but the program keeps going?    the stack trace shows the connection pool attempting to reset a connection, and gevent is rejecting a context switch.  This would suggest the connection pool reset is occurring during Python garbage collection which is run in a separate thread (this would be if you just see stack traces in logs but nothing actually interrupting the program).

the solution would be to make sure connections are explicitly returned to the pool without being garbage collected.

**Reply by andrii-korotkov-verkada** [reply]
I think they are interruptive to the main thread, i.e. tests hang and pods in production stop processing traffic and eventually die.

**Reply by andrii-korotkov-verkada** [reply]
> the solution would be to make sure connections are explicitly returned to the pool without being garbage collected.

What are some good practices for that? Are there some ways to detect when connections are not explicitly returned to the pool?

**Reply by zzzeek** [reply]
well we try to encourage context managers for connections and sessions in modern use.   we've had issues with our context managers regarding cases like greenlets or async coroutines being cancelled or interrupted though we should be pretty good on that now since there are still exceptions that we can catch and we should have those all under control in modern SQLAlchemy versions (including recent 1.4 as you are using).    your stack trace seems to start right at the "finalize" method which indicates this is a GC'ed connection.

mostly you'd want to see if there was some precipitating error before this stack trace occurred.  most of our async / greenlet failures seem to occur in the "return to the pool" stage because it happens after some error condition has occurred and the efforts to resolve things are dealing with a connection that's no longer usable or a context that's stuck in cleanup.

**Reply by andrii-korotkov-verkada** [reply, marked as answer]
Looks like the issue was related to some sessions/queries leaking out of the context. Fixing those leaks seems to fix the issue.
